File: Week 1/pset1/greedy.c
/*
----------------------------------------------------------
|     University of Amsterdam: Minor Programmeren        |
|     Problem Set 1: Greedy                              |
|     Created by Davy Boekhout on 6/09/2017              |
|                                                        |
|     Calculates the minimal amount of coins that        |
|     are needed to give back change                     |
----------------------------------------------------------
*/

#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    float change = 0;
    // Get non-negative change from the user
    while (change <= 0)
    {
        printf("O hai! How much change is owed?\n");
        change = get_float();
    }

    int change_owed = round(change * 100);
    int count = 0;

    while (change_owed != 0)
    {
        if (change_owed > 24)
        {
            count++;
            change_owed = change_owed - 25;
        }
        else if (change_owed > 9)
        {
            count++;
            change_owed = change_owed - 10;
        }
        else if (change_owed > 4)
        {
            count++;
            change_owed = change_owed - 5;
        }
        else
        {
            count++;
            change_owed = change_owed - 1;
        }
    }
    printf("%d\n", count);
}

File: Week 6/pset6/greedy.py
#!/usr/bin/env python3
# ----------------------------------------------------------
# |     University of Amsterdam: Minor Programmeren        |
# |     Problem Set 6: Greedy (less)                       |
# |     Created by Davy Boekhout on 09/10/2017             |
# |                                                        |
# |     Calculates the minimal amount of coins needed      |
# |     to give back a given amount of change.             |
# ----------------------------------------------------------


def get_positive_float():
    """Take input from the user and return a positive floating number"""
    try:
        f = float(input("How much change is owed?\n"))
        if f <= 0.0:
            raise ValueError
    # in case user inputs text or a negative number
    except ValueError:
        return
    else:
        return round(f * 100)


def get_change(change):
    """Returns the minimal amount of coins needed to give back change"""
    coins = 0

    # test how many times change_owed goes into nickel, dime and cent
    while change != 0:
        if change > 24:
            coins += 1
            change -= 25
        elif change > 9:
            coins += 1
            change -= 10
        elif change > 4:
            coins += 1
            change -= 5
        else:
            coins += 1
            change -= 1
    return coins


def main():
    print("O hai! ", end="")

    # validate input
    change_owed = None
    while change_owed == None:
        change_owed = get_positive_float()

    # get the minimal amount of coins needed
    coin_amount = get_change(change_owed)
    print("{}".format(coin_amount))


if __name__ == '__main__':
    main()
